crates/matrix-sdk-crypto/src/types/room_history.rs

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//! Types for sharing encrypted room history, per MSC4268
+
+use std::fmt::Debug;
+
+use ruma::{DeviceKeyAlgorithm, OwnedRoomId};
+use serde::{Deserialize, Serialize};
+use vodozemac::{megolm::ExportedSessionKey, Curve25519PublicKey};
+
+#[cfg(doc)]
+use crate::olm::InboundGroupSession;
+use crate::{
+    olm::ExportedRoomKey,
+    types::{
+        deserialize_curve_key, events::room_key_withheld::RoomKeyWithheldContent,
+        serialize_curve_key, EventEncryptionAlgorithm, SigningKeys,
+    },
+};
+
+/// A bundle of room keys, for sharing encrypted room history.
+#[derive(Deserialize, Serialize, Debug, Default)]
+pub struct RoomKeyBundle {
+    /// Keys that we are sharing with the recipient.
+    pub room_keys: Vec<SharedRoomKey>,
+
+    /// Keys that we are *not* sharing with the recipient.
+    pub withheld: Vec<RoomKeyWithheldContent>,
+}
+
+/// An [`InboundGroupSession`] for sharing as part of the key bundle.
```

Comment:
We should probably describe how this differs from other 4 different room key types we have.

Would be neat to document the dangers of accepting such a room key and that they need to be treated somewhat specially, i.e. they will never be fully trusted like the room keys we receive directly from the creator.

Reply:
I've added a bunch of documentation which I hope will help, in 27bfeb8